### PR TITLE
get timestamps from watcher on invalidate

### LINKED ISSUE #5970: [Bug/Feature] Module file timestamps are only recorded on change events.
**Do you want to request a *feature* or report a *bug*?**
Both-ish. I think it's a bug, but could be a feature.

**What is the current behavior?**
Webpack is only recording file timestamps for modules after changing a file. This is bad because it causes full rebuilds when dynamically adding to a compiler.

**If the current behavior is a bug, please provide the steps to reproduce.**
I created a [demo repository](https://github.com/gdborton/webpack-build-as-needed).

```
git clone git@github.com:gdborton/webpack-build-as-needed.git
cd webpack-build-as-needed
npm install
node server
```

In your console you should see that `file1.bundle.js` was compiled.
```
building: multi /Users/gary_borton/code/webpack-filetimestamp/src/file1.bundle.js
Listening on port: 3000
building: /Users/gary_borton/code/webpack-filetimestamp/src/file1.bundle.js

webpack: Compiled successfully.
```

If you request `http://localhost:3000/dist/file2.bundle.js`, we'll add it to the compiler, and invalidate. You'll see a message about that, and then you'll see that `file1.bundle.js` was compiled again, despite having no changes.

```
Found a request for an unbuilt bundle "file2.bundle", adding to compiler: "myFancyCompiler"
webpack: Compiling...
building: /Users/gary_borton/code/webpack-filetimestamp/src/file1.bundle.js
building: multi /Users/gary_borton/code/webpack-filetimestamp/src/file2.bundle.js
webpack: wait until bundle finished: /dist/file2.bundle.js
building: /Users/gary_borton/code/webpack-filetimestamp/src/file2.bundle.js
```

**What is the expected behavior?**
Since `file1.bundle.js` has already been compiled successfully and hasn't changed, it shouldn't need to be recompiled.

**If this is a feature request, what is motivation or use case for changing the behavior?**
Performance! I think Airbnb (and others) could see huge dev performance benefits by only compiling/recompiling assets as needed.

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**
This isn't really relevant.

cc: @sokra @TheLarkInn @ljharb @iancmyers

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
no

**If relevant, link to documentation update:**
N/A

**Summary**
fixes #5970

**Does this PR introduce a breaking change?**
yes

**Other information**

### CODE REVIEW

File: lib/WatchIgnorePlugin.js [hunk 1]
```diff
@@ -49,5 +49,24 @@ class IgnoringWatchFileSystem {
...
 			callback(err, filesModified, dirsModified, missingModified, fileTimestamps, dirTimestamps);
 		}, callbackUndelayed);
+
+		return {
+			close: () => watcher.close(),
+			pause: () => watcher.pause(),
+			getContextTimestamps: () => {
+				const dirTimestamps = watcher.getContextTimestamps();
+				ignoredDirs.forEach(path => {
+					dirTimestamps[path] = 1;
```
Comment (ooflorent) [inline lib/WatchIgnorePlugin.js]: Using a `Set` would prevent deopt

File: lib/Watching.js [hunk 2]
```diff
@@ -121,6 +121,14 @@ class Watching {
 		if(callback) {
 			this.callbacks.push(callback);
 		}
+		if(this.watcher) {
```
Comment (lusarz) [inline lib/Watching.js]: Is it possible that `this.watcher` may be falsy ?

File: lib/WatchIgnorePlugin.js (hunk 1 above)
Comment (sokra) [reply lib/WatchIgnorePlugin.js]: Good idea to refactor this. -> separate PR.

Comment (TheLarkInn) [reply lib/WatchIgnorePlugin.js]: Created https://github.com/webpack/webpack/issues/6234 for the refactor.

Comment (sokra) [reply lib/WatchIgnorePlugin.js]: Note: It should be a `Map`

**Pull request merged**